app.py

```diff
@@ -516,6 +516,70 @@ def dashboard():
 
     return render_template('private/dashboard.html', current_user=current_user, resources=resources)
 
+# Name approval dashboard
+@app.route('/approve_names', methods=['GET', 'POST'])
+@login_required
+def approve_names():
+    message = ''
+    error = ''
+    connection = db.engine
+    if current_user.gm_level != 9:
+        abort(403)
+        return
+    
+    if request.method == 'POST':
+
+        pending_name = request.form['pending_name']
+        approved = int(request.form['approved'])
+
+        # Character name approvals
+        if "char_id" in request.form and "pet_id" not in request.form:
+            char_id = request.form['char_id']
+            if approved > 0:
+                query = "UPDATE charinfo SET name = %s, pending_name = '' WHERE id = %s"
+                try:
+                    connection.execute(query, (pending_name, char_id))
+                except BaseException as error:
+                    error = "Error running query. Message: {}".format(error)
+                else:
+                    message = 'Character name {} approved'.format(pending_name)
+            else:
+                query = "UPDATE charinfo SET needs_rename = true WHERE id = %s"
+                try:
+                    connection.execute(query, (char_id))
+                except BaseException as error:
+                    error = "Error running query. Message: {}".format(error)
+                else:
+                    message = 'Character {} marked for rename'.format(pending_name)
+        # Pet name approvals
+        elif "pet_id" in request.form and "char_id" not in request.form:
+            pet_id = request.form['pet_id']
+            print(pet_id)
+            print(approved)
+            if approved > 0:
+                query = "UPDATE pet_names SET approved = 2 WHERE id = %s"
+                try:
+                    connection.execute(query, (pet_id))
+                except BaseException as error:
+                    error = "Error running query. Message: {}".format(error)
+                else:
+                    message = 'Pet name {} approved'.format(pending_name)
+            else:
+                query = "UPDATE pet_names SET approved = 0 WHERE id = %s"
```

Comment:
The server itself does not seem to support rejecting pet names yet. Unless I am missing something it only supports pending approval (1) and approved (2) at the moment, see [relevant code](https://github.com/DarkflameUniverse/DarkflameServer/blob/0a7d9c46eba551b5e2f908caf9c689f4139bffb4/dGame/dComponents/PetComponent.cpp#L1205-L1210).

Would it make sense to omit rejecting pet names for now, until that has been implemented server-side?
Otherwise it might cause issues in the future when it is implemented for the server but they chose a different implementation, other than using value 0 to represent renaming required.

Reply:
The server does support rejecting names. Changing the approved value to 0 rejects the name, and forces the user to rename the pet when they bring it out next. 

However, that being said, it does NOT support forcing a user to rename a pet that's already been approved. To accomplish this, you need to set a 0 in the `pet_names` table AND set a 0 on the `m` tag in the respective `pet` tag in the user's xml data.

Reply:
> The server itself does not seem to support rejecting pet names yet. Unless I am missing something it only supports pending approval (1) and approved (2) at the moment, see [relevant code](https://github.com/DarkflameUniverse/DarkflameServer/blob/0a7d9c46eba551b5e2f908caf9c689f4139bffb4/dGame/dComponents/PetComponent.cpp#L1205-L1210).
> 
> Would it make sense to omit rejecting pet names for now, until that has been implemented server-side? Otherwise it might cause issues in the future when it is implemented for the server but they chose a different implementation, other than using value 0 to represent renaming required.


The server does support rejecting names. Changing the approved value to 0 rejects the name, and forces the user to rename the pet when they bring it out next. 

However, that being said, it does NOT support forcing a user to rename a pet that's already been approved. To accomplish this, you need to set a 0 in the `pet_names` table AND set a 0 on the `m` tag in the respective `pet` tag in the user's xml data.

Reply:
Thank you for the clarification!

> However, that being said, it does NOT support forcing a user to rename a pet that's already been approved.

This might not be an issue (for now) because if I understand these changes correctly, only unapproved pet names are displayed. So unless you are manually editing the database to change an already approved name back to unapproved, this situation should not occur.

(Though @Lego6245 can probably answer that better.)

Reply:
Correct on all counts.

While there is no reference in server code, in client code, it handles the state of '0' correctly.

Only pending names are shown. There is no way to re-force a name to be approved (currently).